Harmony/Tools/AccessTools.cs

```diff
@@ -937,6 +937,102 @@ public static FieldRef<F> StaticFieldRefAccess<F>(FieldInfo fieldInfo)
 			return (FieldRef<F>)dm.Generate().CreateDelegate(typeof(FieldRef<F>));
 		}
 
+		/// <summary>Creates a delegate to a given method</summary>
+		/// <typeparam name="DelegateType"></typeparam>
+		/// <param name="method">The method to create a delegate from.</param>
+		/// <param name="instance">
+		/// Only applies for instance methods. If <c>null</c> (default), returned delegate is an open (a.k.a. unbound) instance delegate
+		/// where an instance is supplied as the first argument to the delegate invocation; else, delegate is a closed (a.k.a. bound)
+		/// instance delegate where the delegate invocation always applies to the given <paramref name="instance"/>.
+		/// </param>
+		/// <param name="virtualCall">
+		/// Only applies for instance methods. If <c>true</c> (default), invocation of the delegate calls the instance method virtually;
+		/// else, invocation of the delegate calls the exact instance method (this is useful for calling base class methods).
+		/// </param>
+		/// <returns>A delegate of given <typeparamref name="DelegateType"/> to given <paramref name="method"/></returns>
+		/// <remarks>
+		/// <para>
+		/// Delegate invocation is more performant and more convenient to use than <see cref="MethodBase.Invoke(object, object[])"/>
+		/// at a one-time setup cost.
+		/// </para>
+		/// <para>
+		/// Works for both type of static and instance methods, both open and closed (a.k.a. unbound and bound) instance methods,
+		/// and both class and struct methods. Currently, open instance methods where the delegate type specifies an interface instance type
+		/// are not supported.
+		/// </para>
+		/// </remarks>
+		public static DelegateType MethodDelegate<DelegateType>(MethodInfo method, object instance = null, bool virtualCall = true)
+			where DelegateType : Delegate
+		{
+			if (method is null)
+				throw new ArgumentNullException(nameof(method));
+
+			var delegateType = typeof(DelegateType);
+			if (method.IsStatic)
```

Comment:
Using `if() { return }` without `else` makes the whole construct easier to read with less indention

Reply:
I can remove that `else` but it wouldn't reduce any indentation. Only removing `else` for the `if (instance is null)` would do that, but that's arguably less readable, since it's like (but not exactly):

```cs
if (instance is null)
{
  if (virtualCall)
    ...
  else
    ...
}
else
{
  if (virtualCall)
    ...
  else
    ...
}
```

Reply:
How about this structure:
```
if (static) 
{
   ...
   return
}

if (instance == null)
{
   if (virtual)
   {
       ...
       return
   }
   ...
   return
}

if (virtual) return
...
return
```

Reply:
And maybe switch so instance==null is last

Reply:
Well this where the "but not exactly" comes into play. I'll change it to this:

```cs
if (static)
{
   return ...
}

if (instance is null)
{
   if (virtualCall && !declaringType.IsValueType)
   {
     return ...
   }

   ...
   return ...
}

if (virtualCall)
{
  return ...
}

...
return ...
```

Reply:
Looks fine to me